okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java

```diff
@@ -273,7 +273,11 @@ private void scheduleExecution(Runnable execution, Duration backoff) {
         return new QosException.Visitor<Void>() {
             @Override
             public Void visit(QosException.Throttle exception) {
-                Optional<Duration> nonAdvertizedBackoff = backoffStrategy.nextBackoff();
+                if (!retryStrategy.shouldRetry()) {
+                    throw exception;
```

Comment:
Don't remember 100% of the details, but I believe you should only throw IOException within the okhttp callstack. Maybe an alternative: throw a more special SafeIoException below?

Reply:
Instead of introducing a new `SafeIoException` for the user to handle, how about we return a 404 `Response`? This does not introduce any additional API since the caller already gets back a `Response`. The caller just needs to check the status code and throw a `QosException` if it's 404.

Reply:
@jonsyu1 not sure I understand what you're proposing - surely it's possible we could get a real 404 and wouldn't want that to be turned into a 429..? Would a real 404 show up differently somehow?

Reply:
Sorry, I don't know why I typed 404 when I meant to type 429. I updated my comment.

Reply:
I believe it would be better to pass an exception out via `callback.onFailure(call, ...)` rather than throwing at this point in the code. I think that would make sure that the exception bubbles up in the right way up in https://github.com/palantir/conjure-java-runtime/blob/266a35b5550ff4d50cd69e53e645aa6b609f6723/okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java#L119

On the other hand I don't know this code at all well, so could be talking nonsense.

I feel it would be worth getting to the point where the tests pass with your initial refactors (the `BackoffStrategy` -> `RetryStrategy` piece) but then catching up with someone on the @palantir/infrastructure team in a higher bandwidth setting to figure out the correct approach here.

Reply:
What I'm trying to suggest above is passing a 429 via `callback.onResponse(call, response)` because this is easier for the caller to handle.

For retrofit clients, this means returning a `Call` whose `Response.code()` returns 429. This is more intuitive than catching `IOException` and checking if it is a special `SafeIoException` because it is already part of retrofit's normal API.

For feign clients, this means throwing a `FeignException` whose `status()` returns 429. This is more intuitive than checking the `FeignException`'s cause for a special `SafeIoException` because it is already part of feign's normal API.

In its current form, the `Response` is turned into a `QosException` and the visitor does not get access to the `Response`. We could have the `qosHandler` return the `Response` in addition to the `QosException` but that seems to leak the `Response` from the `ResponseHandler` abstraction. Alternatively, we could construct a new `ResponseHandler` that uses the `qosHandler` to turn the `Response` into a `QosException` and then invokes the `callback.onResponse` with the `Response`.

I'm happy to iterate on this over VTC or slack.